File: backend/Infrastructure/Services/WeatherService.cs
using Core.Interfaces;
using Core.Models;
using Infrastructure.Helpers;
using Microsoft.Extensions.Configuration;
using Newtonsoft.Json;

namespace Infrastructure.Services;

public class WeatherService
{
    private readonly HttpClient _httpClient;
    private readonly string _weatherUrl;
    private readonly string _urlVariables;
    private readonly IUnitOfWork _unitOfWork;
    private readonly AesEncryptionService _encryptionService;

    public WeatherService(HttpClient httpClient, IConfiguration configuration, IUnitOfWork unitOfWork, AesEncryptionService encryptionService)
    {
        _httpClient = httpClient;
        _weatherUrl = configuration["WeatherInfo:URL"];
        _urlVariables = configuration["WeatherInfo:UrlVariables"];
        _unitOfWork = unitOfWork;
        _encryptionService = encryptionService;
    }

    public async Task<WeatherData?> GetWeatherDataAsync(string city)
    {
        var apiKey = await _unitOfWork.ApiKeys.GetKeyByProviderAsync("VisualCrossing");

        if (apiKey == null)
            throw new Exception("API Key not found.");

        apiKey.Key = _encryptionService.Decrypt(apiKey.Key);

        var apiUrl = $"{_weatherUrl}/{city}?key={apiKey.Key}{_urlVariables}";
        var response = await _httpClient.GetStringAsync(apiUrl);

        var apiResponse = JsonConvert.DeserializeObject<ApiResponse>(response);
        var weatherApiData = apiResponse?.Days?.FirstOrDefault();

        if (weatherApiData != null)
        {
            return new WeatherData
            {
                Location = city,
                Temperature = weatherApiData.Temp,
                TempMax = weatherApiData.TempMax,
                TempMin = weatherApiData.TempMin,
                Humidity = weatherApiData.Humidity,
                PrecipProb = weatherApiData.PrecipProb,
                Sunrise = weatherApiData.Sunrise,
                Sunset = weatherApiData.Sunset,
                Description = weatherApiData.Description,
                Pressure = weatherApiData.Pressure,
                Icon = weatherApiData.Icon
            };
        }

        return null;
    }

    public async Task<List<WeatherData>> GetForecastWeatherDataAsync(string city, string date1, string date2)
    {
        var apiKey = await _unitOfWork.ApiKeys.GetKeyByProviderAsync("VisualCrossing");

        if (apiKey is null)
            throw new Exception("API Key not found.");

        apiKey.Key = _encryptionService.Decrypt(apiKey.Key);

        var apiUrl = $"{_weatherUrl}/{city}/{date1}/{date2}?key={apiKey.Key}{_urlVariables}";
        var response = await _httpClient.GetStringAsync(apiUrl);

        var apiResponse = JsonConvert.DeserializeObject<ApiResponse>(response);

        if (apiResponse?.Days != null && apiResponse.Days.Any())
        {
            var weatherList = apiResponse.Days.Select(day => new WeatherData
            {
                Datetime = day.DateTime,
                Location = city,
                Temperature = day.Temp,
                TempMax = day.TempMax,
                TempMin = day.TempMin,
                Humidity = day.Humidity,
                PrecipProb = day.PrecipProb,
                Sunrise = day.Sunrise,
                Sunset = day.Sunset,
                Description = day.Description,
                Pressure = day.Pressure,
                Icon = day.Icon
            }).ToList();

            return weatherList;
        }

        return new List<WeatherData>();
    }
}

File: frontend/src/app/services/weather.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable, throwError } from 'rxjs';
import { Weather } from '../models/weather.model';
import { map, catchError } from 'rxjs/operators';
import { GLOBAL } from '../configuration/configuration.global';

@Injectable({
  providedIn: 'root',
})
export class WeatherService {
  constructor(private http: HttpClient) {}

  // Method for obtaining weather information by city
  getWeatherInfoByCity(city: string): Observable<Weather> {
    const url = `${GLOBAL.apiWeatherUrl}/${city}`;
    console.log('urL: ' + url);
    return this.http.get<any>(url).pipe(
      map((response) => {
        if (!response || !response.location) {
          throw new Error('Datos no válidos de la API');
        }

        const weather: Weather = {
          datetime: new Date().toISOString().split('T')[0],
          location: response.location,
          temperature: response.temperature,
          tempmax: response.tempMax,
          tempmin: response.tempMin,
          humidity: response.humidity,
          precipprob: response.precipProb,
          sunrise: new Date(`1970-01-01T${response.sunrise}Z`),
          sunset: new Date(`1970-01-01T${response.sunset}Z`),
          description: response.description,
          pressure: response.pressure,
          icon: response.icon,
        };

        return weather;
      }),
      catchError((error) => {
        console.error('Error al obtener los datos del clima:', error);
        return throwError(() => new Error('No se pudo obtener el clima.'));
      })
    );
  }

  // Method for obtaining weather forecast information by city for the next 3 days
  getForecastInfoByCity(
    city: string,
    startDate: string,
    endDate: string
  ): Observable<Weather[]> {
    const url = `${GLOBAL.apiWeatherUrl}/forecast/${city}?date1=${startDate}&date2=${endDate}`;
    console.log('URL: ' + url);

    return this.http.get<any>(url).pipe(
      map((response) => {
        console.log('Respuesta de la API:', response);

        if (!Array.isArray(response) || response.length !== 3) {
          throw new Error(
            'Datos no válidos de la API o no hay 3 días disponibles'
          );
        }

        const weatherArray: Weather[] = response.map((day: any) => ({
          datetime: day.datetime,
          location: city,
          temperature: day.temperature,
          tempmax: day.tempMax,
          tempmin: day.tempMin,
          humidity: day.humidity,
          precipprob: day.precipProb,
          sunrise: new Date(`1970-01-01T${day.sunrise}Z`),
          sunset: new Date(`1970-01-01T${day.sunset}Z`),
          description: day.description,
          pressure: day.pressure,
          icon: day.icon,
        }));

        return weatherArray;
      }),
      catchError((error) => {
        console.error('Error al obtener los datos del clima:', error);
        return throwError(() => new Error('No se pudo obtener el clima.'));
      })
    );
  }
}
